Implement named inference rule for mul

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23435 wip
* #23436 Make torch.view drop all names on named tensors
* #23316 Add names to repr for named tensors
* #23315 Rename tensor.is_named to has_named, expose has_named to python.
* #23237 Enable named tensors for arithmetic, clone, and tensor conversion ops
* #23229 Named inference rule for copy_
* **#23193 Implement named inference rule for mul**

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

Differential Revision: [D16494401](https://our.internmc.facebook.com/intern/diff/D16494401)

zou3519 committed:
Implement named inference rule for mul

Featuring:
- The named inference rule is actually more complicated than it should
be. This is because I'd like to "swap out" named tensor rules (I'm
experimenting with two of them) easily. In the future when things are
more stable I will return to the callsites and clean up the API.

Test Plan:
- [namedtensor ci]

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- The named inference rule is actually more complicated than it should
be. This is because I'd like to "swap out" named tensor rules (I'm
experimenting with two of them) easily. In the future when things are
more stable I will return to the callsites and clean up the API.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- The named inference rule is actually more complicated than it should
be. This is because I'd like to "swap out" named tensor rules (I'm
experimenting with two of them) easily. In the future when things are
more stable I will return to the callsites and clean up the API.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- The named inference rule is actually more complicated than it should
be. This is because I'd like to "swap out" named tensor rules (I'm
experimenting with two of them) easily. In the future when things are
more stable I will return to the callsites and clean up the API.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- The named inference rule is actually more complicated than it should
be. This is because I'd like to "swap out" named tensor rules (I'm
experimenting with two of them) easily. In the future when things are
more stable I will return to the callsites and clean up the API.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

colesbury commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -80,6 +83,17 @@ Tensor& mul_out(Tensor& result, const Tensor& self, const Tensor& other) {
     return at::_sparse_mul_out(result, self, other);
   }
   at::assert_no_internal_overlap(result, "mul");
+#ifdef BUILD_NAMEDTENSOR
```
Can you implement this in one place in TensorIterator instead of each of the call sites?

zou3519 replied:
Sure, let me look into it.

zou3519 replied:
Thanks for the great suggestion, @colesbury, it worked out really well

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

nairbv commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -213,6 +213,40 @@ void TensorIterator::allocate_outputs() {
+
+  // propagate names
+  for (int i = 0; i < num_outputs_; i++) {
+    auto& op = operands_[i];
```
Do you happen to know why `Tensor output(int arg=0)` doesn't return a `Tensor&` ? Seems like it would be nice to be able to use that here.

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head

zou3519 committed:
Update on "Implement named inference rule for mul"

Implement named inference rule for mul

Featuring:
- Implemented name propagation directly in TensorIterator. This also
probably enables a lot of other functions but I'll add/enable them in
another PR.
- Cleaned up old callsites that were using TensorIterator since
TensorIterator automatically propagates names.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23193 gh/zou3519/75/head
